modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,280 @@
+
+  return persist_name;
+}
+
+static gboolean
+_check_is_valid_protocol(gchar *protocol_type)
+{
+  gchar* valid_type[] = {"tcp", "ssl", "ws", "wss"};
+  gint i;
+
+  for (i = 0; i < G_N_ELEMENTS(valid_type); ++i)
+      if (strcmp(protocol_type, valid_type[i]) == 0)
+        return TRUE;
+
+  return FALSE;
+}
+
+static void
+_set_default_value(LogPipe *s)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (self->host->len == 0)
+    g_string_assign(self->host, DEFAULT_HOST);
+
+  if (self->port == -1)
+    self->port = DEFAULT_PORT;
+
+  if (self->keepalive == -1)
+    self->keepalive = DEFAULT_KEEPALIVE;
+
+  if (self->qos == -1)
+    self->qos = DEFAULT_QOS;
+
+  if (self->protocol_type->len == 0)
+    g_string_assign(self->protocol_type, DEFAULT_PROTOCOL_TYPE);
+  
+  if (self->message == NULL)
+  {
+    self->message = log_template_new(cfg, NULL);
+    log_template_compile(self->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+  }
+
+  log_template_options_init(&self->template_options, cfg);
+}
+
+static gboolean
+_dd_init(LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  
+  _set_default_value(d);
+
+  if (!log_threaded_dest_driver_init_method(d))
+    {
+      msg_error("ERROR, while init threaded dest.");
+      return FALSE;
+    }
+
+  if (self->topic->len == 0)
+    {
+      msg_error("ERROR, topic must be set!"); 
+      return FALSE;
+    }
+  
+
+  if (self->super.batch_lines != -1)
+    {
+      msg_error("ERROR, can not be set batch_lines() in mqtt destination!");
+      return FALSE;
+    }
+
+  if (self->super.batch_timeout != -1)
+    {
+      msg_error("ERROR, can not be set batch_timeout() in mqtt destination!");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+gboolean
+_dd_deinit(LogPipe *s)
+{
+  // MQTTDestinationDriver *self = (MQTTDestinationDriver *)s;
+
+  return log_threaded_dest_driver_deinit_method(s);
+}
+
+static void
+_dd_free(LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  g_string_free(self->host, TRUE);
+  g_string_free(self->topic, TRUE);
+  g_string_free(self->protocol_type, TRUE);
+  
+  log_template_options_destroy(&self->template_options);
+  log_template_unref(self->message);
+
+  log_threaded_dest_driver_free(d);
+}
+
+static void
+_mqtt_destination_driver_basic_values(MQTTDestinationDriver *driver)
+{
+  driver->port          = -1;
+  driver->clean_session = DEFAULT_CLEAN_SESSION;
+  driver->keepalive     = -1;
+  driver->qos           = -1;
+  // need to free later
+  driver->host          = g_string_new("");
+  driver->topic         = g_string_new("");
+  driver->protocol_type = g_string_new("");
+
+  log_template_options_defaults(&driver->template_options);
+
+}
+
+LogDriver *
+mqtt_destination_dd_new(GlobalConfig *cfg)
+{
+  MQTTDestinationDriver *self = g_new0(MQTTDestinationDriver, 1);
+
+  _mqtt_destination_driver_basic_values(self);
```

Comment:
this should follow the basic initialization of the class (at the very least after init_instance), maybe could even be uninlined.

Also the name "basic_values" should probably refer to "defaults".

Reply:
solved by this commit: MQTT/destination: remove basic value. using default value insted of basic value
(commit hash: 67545ae92b00887a6b1d9ee96c56d8c8f4c10203 )